src/highlights/highlight_grid.ts

```diff
@@ -313,6 +269,61 @@ export class HighlightGrid {
 
         return currMaxCol;
     }
+
+    private rangesForRow(rowHighlights: Highlight[][], row: number, topLine: number): HighlightRange[] {
+        const line = row + topLine;
+
+        const normalHighlights: Map<number, NormalTextHighlightRange[]> = new Map();
+        const virtualHighlights: VirtualTextHighlightRange[] = [];
+        rowHighlights.forEach((colHighlights, col) => {
+            if (colHighlights.length === 0) {
+                // Should never happen, but defensive
+                return;
+            }
+
+            if (colHighlights.length > 1 || colHighlights[0].virtText) {
+                virtualHighlights.push({
+                    textType: "virtual",
+                    highlights: colHighlights,
+                    line,
+                    col,
+                });
+                return;
+            }
+
+            const colHighlight = colHighlights[0];
+            const existingHighlights = normalHighlights.get(colHighlight.hlId) ?? [];
+            const matchingHighlight = rfind(
```

Comment:
findLast?

Reply:
Sure. I just stole the name from Python, haha

Reply:
I mean `existingHighlights.findLast`

Reply:
Oh, huh.  I couldn't find that before when I went looking for it

Reply:
Ah, ok. `findLast` is only available in ES2023, and we target ES2017. We could change that I guess? Not sure the implications of doing so to be honest